_source/_posts/2021-09-12-spring-boot-spa.md

```diff
@@ -0,0 +1,686 @@
+
+## Spring Boot and CORS
+
+CORS is a protocol that allows for browsers and servers to explicitly allow cross-origin resource sharing. By default, Javascript running in a browser is restrained by a **same-origin policy**. This means that a Javascript application can only make calls to a server that resides on the same domain from which it was loaded. As a starting point, this makes a lot of sense from a security perspective and prevents a lot of abuse. However, this policy would be pretty restrictive if developers were not able to allow exceptions.
+
+CORS allows resource servers to explicitly enable cross-origin requests, to tell the browser what type of requests it allows and from what origins. It is a white-listing scheme in the sense that all cross-origin requests will be denied by the browser unless the server is explicitly configured to allow them. 
+
+Notice that this protocol is mediated between the server and the browser. The client application, the Vue app in this case, doesn't really have to do anything about CORS. From the perspective of the client app, the HTTP request to the resource server will either succeed or fail. If it fails because of CORS, the request will return a `401 (Unauthorized)`.
+
+In this example, the resource server resides on `http://localhost:8082` and the Vue client will be loaded from from the Vue development server at `http://localhost:8080`. Because the port number is part of the domain, these two URLs are considered different origins by the browser. When the Vue app tries to make a request to the Spring Boot resource server, the browser will throw an error unless CORS is properly configured.
+
+Note that not every request will trigger a CORS check. Certain types of simple requests are allowed, particularly requests that only allow form and text content types and the `GET`, `HEAD`, and `POST` request types. Take a look at the [Mozilla developer docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) on this for more info. However, any request with the `Content-Type` header `application-json` will always trigger a CORS check if made to a cross-origin domain.
+
+The first thing the browser will do when a suitable request is made to a resource on a different domain is to send a CORS preflight request. This is actually a totally separate request made by the browser **before** sends the actual request. The CORS preflight request essentially asks the browser if it is configured for CORS, if it will allow the type of request the client wants to send (DELETE, GET, POST, etc.), and if it will allow requests from the specific origin domain. The preflight request has nothing to do with authentication or authorization in the way we normally speak about it. It's just a quick check to see if CORS is configured and if, in theory, the server might allow the request based on the HTTP verb and the client domain.
```

Comment:
```suggestion
The first thing the browser will do when a suitable request is made to a resource on a different domain is to send a CORS preflight request. This is actually a totally separate request made by the browser **before** it sends the actual request. The CORS preflight request essentially asks the browser if it is configured for CORS, if it will allow the type of request the client wants to send (DELETE, GET, POST, etc.), and if it will allow requests from the specific origin domain. The preflight request has nothing to do with authentication or authorization in the way we normally speak about it. It's just a quick check to see if CORS is configured and if, in theory, the server might allow the request based on the HTTP verb and the client domain.
```
missing pronoun